src/Utility/ChatTools.cs

```diff
@@ -0,0 +1,296 @@
+    public async Task<ChatCompletionOptions> CreateCompletionOptionsAsync(string prompt, int maxTools = 3, float minVectorDistance = 0.29f)
+    {
+        if (!CanFilterTools)
+            return CreateCompletionOptions();
+
+        var completionOptions = new ChatCompletionOptions();
+        foreach (var tool in await FindRelatedTools(true, prompt, maxTools, minVectorDistance).ConfigureAwait(false))
+            completionOptions.Tools.Add(tool);
+        return completionOptions;
+    }
+
+    private async Task<IEnumerable<ChatTool>> FindRelatedTools(bool async, string prompt, int maxTools, float minVectorDistance)
+    {
+        if (!CanFilterTools)
+            return _tools;
+
+        return (await FindVectorMatches(async, prompt, maxTools, minVectorDistance).ConfigureAwait(false))
+            .Select(e => ParseToolDefinition(e.Data));
+    }
+
+    private async Task<IEnumerable<VectorbaseEntry>> FindVectorMatches(bool async, string prompt, int maxTools, float minVectorDistance)
+    {
+        var vector = async ?
+            await ToolsUtility.GetEmbeddingAsync(_client, prompt).ConfigureAwait(false) :
+            ToolsUtility.GetEmbedding(_client, prompt);
+
+        lock (_entries)
+        {
+            return ToolsUtility.GetClosestEntries(_entries, maxTools, minVectorDistance, vector);
+        }
+    }
+
+    /// <summary>
+    /// Implicitly converts ChatTools to <see cref="ChatCompletionOptions"/>.
+    /// </summary>
+    /// <param name="tools">The ChatTools instance to convert.</param>
+    public static implicit operator ChatCompletionOptions(ChatTools tools) => tools.CreateCompletionOptions();
+
+    internal string CallLocal(ChatToolCall call)
+    {
+        var arguments = new List<object>();
+        if (call.FunctionArguments != null)
+        {
+            ToolsUtility.ParseFunctionCallArgs(call.FunctionArguments, out arguments);
+        }
+        return CallLocal(call.FunctionName, [.. arguments]);
+    }
+
+    private string CallLocal(string name, object[] arguments)
+    {
+        if (!_methods.TryGetValue(name, out MethodInfo method))
+            return $"I don't have a tool called {name}";
+
+        object result = method.Invoke(null, arguments);
+        return result?.ToString() ?? string.Empty;
+    }
+
+    internal async Task<string> CallMcpAsync(ChatToolCall call)
+    {
+        if (!_mcpMethods.TryGetValue(call.FunctionName, out var method))
+            throw new NotImplementedException($"MCP tool {call.FunctionName} not found.");
+
+#if !NETSTANDARD2_0
+        var actualFunctionName = call.FunctionName.Split(ToolsUtility.McpToolSeparator, 2)[1];
+#else
+        var index = call.FunctionName.IndexOf(ToolsUtility.McpToolSeparator);
+        var actualFunctionName = call.FunctionName.Substring(index + ToolsUtility.McpToolSeparator.Length);
+#endif
+        var result = await method(actualFunctionName, call.FunctionArguments).ConfigureAwait(false);
+        return result.ToString();

```

Comment:
This should probably throw an exception if `result` or `ToString` return `null` explaining that the result of a function cannot be null and the user needs to fix their code.